ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/operators/SelectOperator.java

```diff
@@ -69,34 +69,32 @@ public void open() {
 
   @Override
   public Object next() {
-    row = (TableRow)child.next();
-    if (row == null) {
-      return null;
-    }
-
-    final GenericRow intermediate = PullPhysicalOperatorUtil.getIntermediateRow(
-        row, logicalNode.getAddAdditionalColumnsToIntermediateSchema());
-    final Function<TableRow, Optional<TableRow>> transformRow = row -> transformer.transform(
-        row.key(),
-        intermediate,
-        new PullProcessingContext(row.rowTime()))
-        .map(r -> {
-          if (logicalNode.isWindowed()) {
-            return WindowedRow.of(
-                logicalNode.getIntermediateSchema(),
-                ((WindowedRow) row).windowedKey(),
-                r,
-                row.rowTime());
-          }
-          return Row.of(logicalNode.getIntermediateSchema(), row.key(), r, row.rowTime());
-        });
-
-    Optional<TableRow> result = transformRow.apply(row);
+    Optional<TableRow> result = Optional.empty();
     while (result.equals(Optional.empty())) {
       row = (TableRow)child.next();
       if (row == null) {
         return null;
       }
+
+      final GenericRow intermediate = PullPhysicalOperatorUtil.getIntermediateRow(
```

Comment:
Yes, there was a bug before.

if `result.equals(Optional.empty())` was true, it would try to get a new row and call `result = transformRow.apply(row);`

The subtle issue is that `transformRow` makes reference to `intermediate`, which depends on `row`, so in the event that it tries to get a new row, `transformRow` gives the wrong answer.

I rewrote it so that this process is now within the while loop, including creating `intermediate`.

Reply:
Got it, thanks!